File: model/src/test/java/de/danoeh/antennapod/model/feed/FeedTest.java
package de.danoeh.antennapod.model.feed;

import org.junit.Before;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;

import static de.danoeh.antennapod.model.feed.FeedMother.IMAGE_URL;
import static de.danoeh.antennapod.model.feed.FeedMother.anyFeed;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertThrows;

public class FeedTest {

    private Feed original;
    private Feed changedFeed;

    @Before
    public void setUp() {
        original = anyFeed();
        changedFeed = anyFeed();
    }

    @Test
    public void testUpdateFromOther_feedImageDownloadUrlChanged() {
        changedFeed.setImageUrl("http://example.com/new_picture");
        original.updateFromOther(changedFeed);
        assertEquals(original.getImageUrl(), changedFeed.getImageUrl());
    }

    @Test
    public void testUpdateFromOther_feedImageRemoved() {
        changedFeed.setImageUrl(null);
        original.updateFromOther(changedFeed);
        assertEquals(anyFeed().getImageUrl(), original.getImageUrl());
    }

    @Test
    public void testUpdateFromOther_feedImageAdded() {
        original.setImageUrl(null);
        changedFeed.setImageUrl("http://example.com/new_picture");
        original.updateFromOther(changedFeed);
        assertEquals(original.getImageUrl(), changedFeed.getImageUrl());
    }

    @Test
    public void testSetSortOrder_OnlyIntraFeedSortAllowed() {
        for (SortOrder sortOrder : SortOrder.values()) {
            if (sortOrder.scope == SortOrder.Scope.INTRA_FEED) {
                original.setSortOrder(sortOrder); // should be okay
            } else {
                assertThrows(IllegalArgumentException.class, () -> original.setSortOrder(sortOrder));
            }
        }
    }

    @Test
    public void testSetSortOrder_NullAllowed() {
        original.setSortOrder(null); // should be okay
    }

    @Test
    public void equalsSameIdIsEqual() {
        Feed a = anyFeed();
        Feed b = anyFeed();
        a.setId(1);
        b.setId(1);
        assertEquals(a, b);
    }

    @Test
    public void equalsDifferentIdNotEqual() {
        Feed a = anyFeed();
        Feed b = anyFeed();
        a.setId(1);
        b.setId(2);
        assertNotEquals(a, b);
    }

    @Test
    public void equalsDifferentClassNotEqual() {
        Feed a = anyFeed();
        assertNotEquals(a, "not a feed");
    }

    @Test
    public void hashCodeMatchesForSameId() {
        Feed a = anyFeed();
        Feed b = anyFeed();
        a.setId(1);
        b.setId(1);
        assertEquals(a.hashCode(), b.hashCode());
    }

    @Test
    public void getIdentifyingValuePrefersFeedIdentifier() {
        Feed feed = createFeed("feed-identifier", "http://example.com/download", "title", null, "http://example.com/link");
        assertEquals("feed-identifier", feed.getIdentifyingValue());
    }

    @Test
    public void getIdentifyingValueFallsBackToDownloadUrl() {
        Feed feed = createFeed(null, "http://example.com/download", "title", null, "http://example.com/link");
        assertEquals("http://example.com/download", feed.getIdentifyingValue());
    }

    @Test
    public void getIdentifyingValueFallsBackToTitle() {
        Feed feed = createFeed(null, null, "title", null, "http://example.com/link");
        assertEquals("title", feed.getIdentifyingValue());
    }

    @Test
    public void getIdentifyingValueFallsBackToLink() {
        Feed feed = createFeed(null, null, null, null, "http://example.com/link");
        assertEquals("http://example.com/link", feed.getIdentifyingValue());
    }

    @Test
    public void getHumanReadableIdentifierPrefersCustomTitle() {
        Feed feed = createFeed(null, "http://example.com/download", "title", "custom title", "http://example.com/link");
        assertEquals("custom title", feed.getHumanReadableIdentifier());
    }

    @Test
    public void getHumanReadableIdentifierFallsBackToFeedTitle() {
        Feed feed = createFeed(null, "http://example.com/download", "title", null, "http://example.com/link");
        assertEquals("title", feed.getHumanReadableIdentifier());
    }

    @Test
    public void getHumanReadableIdentifierFallsBackToDownloadUrl() {
        Feed feed = createFeed(null, "http://example.com/download", null, null, "http://example.com/link");
        assertEquals("http://example.com/download", feed.getHumanReadableIdentifier());
    }

    @Test
    public void constructorParsesFundingLinks() {
        Feed feed = new Feed(1, null, "title", "http://example.com/link", "description",
                "http://example.com/pay", "author", "en", null, null, IMAGE_URL, null,
                "http://example.com/feed", 0);
        ArrayList<FeedFunding> paymentLinks = feed.getPaymentLinks();
        assertEquals(1, paymentLinks.size());
        assertEquals("http://example.com/pay", paymentLinks.get(0).url);
    }

    @Test
    public void serializationRoundTripPreservesIdTitleAndFunding() throws Exception {
        Feed feed = anyFeed();
        feed.setId(99);
        feed.setCustomTitle("My Custom Title");

        ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
        try (ObjectOutputStream out = new ObjectOutputStream(byteStream)) {
            out.writeObject(feed);
        }
        Feed deserialized;
        try (ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(byteStream.toByteArray()))) {
            deserialized = (Feed) in.readObject();
        }

        assertEquals(99, deserialized.getId());
        assertEquals(feed.getFeedTitle(), deserialized.getFeedTitle());
        assertEquals("My Custom Title", deserialized.getCustomTitle());
        assertEquals(feed.getPaymentLinks(), deserialized.getPaymentLinks());
    }

    private Feed createFeed(String feedIdentifier, String downloadUrl, String title, String customTitle,
                             String link) {
        return new Feed(1, null, title, customTitle, link, null, null, null, null, null,
                feedIdentifier, null, null, downloadUrl, 0, false, null, null, null, false, Feed.STATE_SUBSCRIBED);
    }
}

File: model/src/test/java/de/danoeh/antennapod/model/feed/FeedTest.kt
package de.danoeh.antennapod.model.feed

import de.danoeh.antennapod.model.feed.FeedMother.IMAGE_URL
import de.danoeh.antennapod.model.feed.FeedMother.anyFeed
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertThrows
import org.junit.Before
import org.junit.Test

class FeedTest {

    private lateinit var original: Feed
    private lateinit var changedFeed: Feed

    @Before
    fun setUp() {
        original = anyFeed()
        changedFeed = anyFeed()
    }

    @Test
    fun testUpdateFromOther_feedImageDownloadUrlChanged() {
        changedFeed.imageUrl = "http://example.com/new_picture"
        original.updateFromOther(changedFeed)
        assertEquals(original.imageUrl, changedFeed.imageUrl)
    }

    @Test
    fun testUpdateFromOther_feedImageRemoved() {
        changedFeed.imageUrl = null
        original.updateFromOther(changedFeed)
        assertEquals(anyFeed().imageUrl, original.imageUrl)
    }

    @Test
    fun testUpdateFromOther_feedImageAdded() {
        original.imageUrl = null
        changedFeed.imageUrl = "http://example.com/new_picture"
        original.updateFromOther(changedFeed)
        assertEquals(original.imageUrl, changedFeed.imageUrl)
    }

    @Test
    fun testSetSortOrder_OnlyIntraFeedSortAllowed() {
        for (sortOrder in SortOrder.entries) {
            if (sortOrder.scope == SortOrder.Scope.INTRA_FEED) {
                original.sortOrder = sortOrder // should be okay
            } else {
                assertThrows(IllegalArgumentException::class.java) { original.sortOrder = sortOrder }
            }
        }
    }

    @Test
    fun testSetSortOrder_NullAllowed() {
        original.sortOrder = null // should be okay
    }

    @Test
    fun equalsSameIdIsEqual() {
        val a = anyFeed()
        val b = anyFeed()
        a.id = 1
        b.id = 1
        assertEquals(a, b)
    }

    @Test
    fun equalsDifferentIdNotEqual() {
        val a = anyFeed()
        val b = anyFeed()
        a.id = 1
        b.id = 2
        assertNotEquals(a, b)
    }

    @Test
    fun equalsDifferentClassNotEqual() {
        val a = anyFeed()
        assertNotEquals(a, "not a feed")
    }

    @Test
    fun hashCodeMatchesForSameId() {
        val a = anyFeed()
        val b = anyFeed()
        a.id = 1
        b.id = 1
        assertEquals(a.hashCode(), b.hashCode())
    }

    @Test
    fun getIdentifyingValuePrefersFeedIdentifier() {
        val feed = createFeed(
            "feed-identifier",
            "http://example.com/download",
            "title",
            null,
            "http://example.com/link"
        )
        assertEquals("feed-identifier", feed.getIdentifyingValue())
    }

    @Test
    fun getIdentifyingValueFallsBackToDownloadUrl() {
        val feed = createFeed(null, "http://example.com/download", "title", null, "http://example.com/link")
        assertEquals("http://example.com/download", feed.getIdentifyingValue())
    }

    @Test
    fun getIdentifyingValueFallsBackToTitle() {
        val feed = createFeed(null, null, "title", null, "http://example.com/link")
        assertEquals("title", feed.getIdentifyingValue())
    }

    @Test
    fun getIdentifyingValueFallsBackToLink() {
        val feed = createFeed(null, null, null, null, "http://example.com/link")
        assertEquals("http://example.com/link", feed.getIdentifyingValue())
    }

    @Test
    fun getHumanReadableIdentifierPrefersCustomTitle() {
        val feed = createFeed(
            null,
            "http://example.com/download",
            "title",
            "custom title",
            "http://example.com/link"
        )
        assertEquals("custom title", feed.getHumanReadableIdentifier())
    }

    @Test
    fun getHumanReadableIdentifierFallsBackToFeedTitle() {
        val feed = createFeed(null, "http://example.com/download", "title", null, "http://example.com/link")
        assertEquals("title", feed.getHumanReadableIdentifier())
    }

    @Test
    fun getHumanReadableIdentifierFallsBackToDownloadUrl() {
        val feed = createFeed(null, "http://example.com/download", null, null, "http://example.com/link")
        assertEquals("http://example.com/download", feed.getHumanReadableIdentifier())
    }

    @Test
    fun constructorParsesFundingLinks() {
        val feed = Feed(
            1, null, "title", "http://example.com/link", "description",
            "http://example.com/pay", "author", "en", null, null, IMAGE_URL, null,
            "http://example.com/feed", 0
        )
        // getPaymentLinks() delegates to FeedFunding.extractPaymentLinks, which only returns null
        // for a blank input; "http://example.com/pay" above is non-blank, so this is always non-null.
        val paymentLinks = feed.getPaymentLinks()!!
        assertEquals(1, paymentLinks.size)
        assertEquals("http://example.com/pay", paymentLinks.get(0).url)
    }

    @Test
    fun serializationRoundTripPreservesIdTitleAndFunding() {
        val feed = anyFeed()
        feed.id = 99
        feed.setCustomTitle("My Custom Title")

        val byteStream = ByteArrayOutputStream()
        ObjectOutputStream(byteStream).use { out -> out.writeObject(feed) }
        val deserialized = ObjectInputStream(ByteArrayInputStream(byteStream.toByteArray())).use { input ->
            input.readObject() as Feed
        }

        assertEquals(99L, deserialized.id)
        assertEquals(feed.feedTitle, deserialized.feedTitle)
        assertEquals("My Custom Title", deserialized.getCustomTitle())
        assertEquals(feed.getPaymentLinks(), deserialized.getPaymentLinks())
    }

    private fun createFeed(
        feedIdentifier: String?,
        downloadUrl: String?,
        title: String?,
        customTitle: String?,
        link: String?
    ): Feed {
        return Feed(
            1, null, title, customTitle, link, null, null, null, null, null,
            feedIdentifier, null, null, downloadUrl, 0, false, null, null, null, false, Feed.STATE_SUBSCRIBED
        )
    }
}
